### ISSUE TITLE
How can I implement multi-label classification as shown in the illustration on the homepage?

### ISSUE THREAD

**Issue by ZH-UCAS** [ultralytics/ultralytics#7067]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Thank you for the open-source project. I noticed that the illustrative image for classification on the homepage depicts multi-label classification. However, upon checking https://docs.ultralytics.com/tasks/classify/ and tutorial videos, the probabilities seem to be after softmax. I'm interested in implementing multi-label classification with multi-hot encoding. Is there a guide or documentation available to achieve multi-label classification as depicted in the illustration?
![image](https://github.com/ultralytics/ultralytics/assets/23288916/689e247e-25c9-417b-83fa-df16388ea190)

**Comment by ZH-UCAS** [comment]
Thank you very much for your response. I've studied the existing documentation and code, and I'm planning to attempt modifying the classification loss to BCE (Binary Cross Entropy) and using the sigmoid activation function. I appreciate your guidance.

**Comment by glenn-jocher** [comment]
@ZH-UCAS Sounds like a solid plan! Modifying the loss to BCE and using the sigmoid activation function are indeed the right steps for multi-label classification. Best of luck with your implementation! If you hit roadblocks or have further questions, the community and we are here to help. Happy coding! 🚀

**Comment by hoominchu** [comment]
Hi @ZH-UCAS thanks for this question. I am trying to do the exact same thing. I have the following questions:
- Could you share how you prepared your custom dataset for this?
- Is it possible to share the changes you made to the code to train with this data?

Thanks you!

**Comment by dalmeidalorraine** [comment]
@hoominchu Did you get a response or implement the changes for multi-label classification?

**Comment by SutirthaChakraborty** [comment]
Hi any update on this thread? @hoominchu @glenn-jocher \

**Comment by glenn-jocher** [comment]
Hello @SutirthaChakraborty and @ZH-UCAS,

Thank you for your interest in implementing multi-label classification with YOLOv8! Here’s a detailed guide to help you both get started:

### Preparing Your Custom Dataset

1. **Dataset Structure**: Ensure your dataset is structured correctly. Each image should have a corresponding label file that contains the multi-hot encoded labels.

2. **Annotation Format**: For multi-label classification, you can use a CSV file where each row corresponds to an image and its labels. For example:
    ```csv
    image_path,label1,label2,label3,...
    path/to/image1.jpg,1,0,1,...
    path/to/image2.jpg,0,1,1,...
    ```

3. **YAML Configuration**: Create a dataset configuration YAML file similar to the one used for single-label classification but adapted for multi-label. Here’s an example:
    ```yaml
    path: ../datasets/my_multilabel_dataset
    train: images/train
    val: images/val
    nc: 3  # number of classes
    names: ['class1', 'class2', 'class3']
    ```

### Modifying the Code for Multi-Label Classification

1. **Loss Function**: Change the loss function to Binary Cross Entropy (BCE). You can do this by modifying the training script to use `torch.nn.BCEWithLogitsLoss` instead of `CrossEntropyLoss`.

2. **Activation Function**: Replace the softmax activation with a sigmoid activation in the model’s final layer to handle multi-label outputs.

Here’s a snippet to illustrate these changes:

```python
import torch
import torch.nn as nn
import torch.optim as optim
from ultralytics import YOLO

# Load your model
model = YOLO('yolov8n.yaml')  # or your custom model

# Modify the final layer to use sigmoid activation
model.model[-1] = nn.Sequential(
    nn.Linear(model.model[-1].in_features, model.model[-1].out_features),
    nn.Sigmoid()
)

# Use BCEWithLogitsLoss for multi-label classification
criterion = nn.BCEWithLogitsLoss()

# Example training loop
optimizer = optim.Adam(model.parameters(), lr=0.001)
for epoch in range(num_epochs):
    for images, labels in dataloader:
        optimizer.zero_grad()
        outputs = model(images)
        loss = criterion(outputs, labels)
        loss.backward()
        optimizer.step()
```

### Training the Model

1. **Training Script**: Use the modified training script to train your model on the multi-label dataset. Ensure you pass the correct dataset path and configuration.

```bash
yolo train data=my_multilabel_dataset.yaml model=yolov8n.pt epochs=100 imgsz=640
```

### Additional Resources

For more detailed guidance, you can refer to the [Ultralytics documentation](https://docs.ultralytics.com/help/FAQ/) and the [PyTorch documentation](https://pytorch.org/docs/stable/nn.html#torch.nn.BCEWithLogitsLoss).

If you encounter any issues or have further questions, feel free to ask. The community and we are here to help!

Best of luck with your multi-label classification project! 🚀

---
Feel free to share your progress or any specific issues you encounter along the way. Happy coding! 😊

**Comment by ganganngannn** [comment]
> Hello @SutirthaChakraborty and @ZH-UCAS,
> 
> Thank you for your interest in implementing multi-label classification with YOLOv8! Here’s a detailed guide to help you both get started:
> 
> ### Preparing Your Custom Dataset
> 1. **Dataset Structure**: Ensure your dataset is structured correctly. Each image should have a corresponding label file that contains the multi-hot encoded labels.
> 2. **Annotation Format**: For multi-label classification, you can use a CSV file where each row corresponds to an image and its labels. For example:
>    ```
>    image_path,label1,label2,label3,...
>    path/to/image1.jpg,1,0,1,...
>    path/to/image2.jpg,0,1,1,...
>    ```
> 3. **YAML Configuration**: Create a dataset configuration YAML file similar to the one used for single-label classification but adapted for multi-label. Here’s an example:
>    ```yaml
>    path: ../datasets/my_multilabel_dataset
>    train: images/train
>    val: images/val
>    nc: 3  # number of classes
>    names: ['class1', 'class2', 'class3']
>    ```
> 
> ### Modifying the Code for Multi-Label Classification
> 1. **Loss Function**: Change the loss function to Binary Cross Entropy (BCE). You can do this by modifying the training script to use `torch.nn.BCEWithLogitsLoss` instead of `CrossEntropyLoss`.
> 2. **Activation Function**: Replace the softmax activation with a sigmoid activation in the model’s final layer to handle multi-label outputs.
> 
> Here’s a snippet to illustrate these changes:
> 
> ```python
> import torch
> import torch.nn as nn
> import torch.optim as optim
> from ultralytics import YOLO
> 
> # Load your model
> model = YOLO('yolov8n.yaml')  # or your custom model
> 
> # Modify the final layer to use sigmoid activation
> model.model[-1] = nn.Sequential(
>     nn.Linear(model.model[-1].in_features, model.model[-1].out_features),
>     nn.Sigmoid()
> )
> 
> # Use BCEWithLogitsLoss for multi-label classification
> criterion = nn.BCEWithLogitsLoss()
> 
> # Example training loop
> optimizer = optim.Adam(model.parameters(), lr=0.001)
> for epoch in range(num_epochs):
>     for images, labels in dataloader:
>         optimizer.zero_grad()
>         outputs = model(images)
>         loss = criterion(outputs, labels)
>         loss.backward()
>         optimizer.step()
> ```
> 
> ### Training the Model
> 1. **Training Script**: Use the modified training script to train your model on the multi-label dataset. Ensure you pass the correct dataset path and configuration.
> 
> ```shell
> yolo train data=my_multilabel_dataset.yaml model=yolov8n.pt epochs=100 imgsz=640
> ```
> 
> ### Additional Resources
> For more detailed guidance, you can refer to the [Ultralytics documentation](https://docs.ultralytics.com/help/FAQ/) and the [PyTorch documentation](https://pytorch.org/docs/stable/nn.html#torch.nn.BCEWithLogitsLoss).
> 
> If you encounter any issues or have further questions, feel free to ask. The community and we are here to help!
> 
> Best of luck with your multi-label classification project! 🚀
> 
> Feel free to share your progress or any specific issues you encounter along the way. Happy coding! 😊

How to define dataloader according to yaml file in ultralytics?

**Comment by Y-T-G** [comment]
You could check this branch for multi-label classification 

https://github.com/ultralytics/ultralytics/pull/16015
